sklearn/tree/_classes.py

```diff
@@ -284,36 +309,51 @@ def fit(self, X, y, sample_weight=None, check_input=True):
         elif self.max_features is None:
             max_features = self.n_features_in_
         elif isinstance(self.max_features, numbers.Integral):
+            check_scalar(
+                self.max_features,
+                "max_features",
+                target_type=numbers.Integral,
+                min_val=1,
+                max_val=self.n_features_in_,
+            )
             max_features = self.max_features
         else:  # float
+            check_scalar(
+                self.max_features,
+                "max_features",
+                target_type=numbers.Real,
+                min_val=0.0,
+                max_val=1.0,
+                include_boundaries="right",
```

Comment:
Here there is a subtle ambiguity for `1` that could be an `Integral` or `Real` whether one pass `1.0` or `1`.

Basically, I would expect:

- `None` allows taking all features.
- `1` as an integer will select one feature.
- `1.0` as a real should not be included.

But I am not really sure what is the best here. Maybe @thomasjpfan or @ogrisel remember how we want to make the validation for this ambiguity.

Reply:
My impression was that `1.0` means "take all the features", which is inline with the current implementation.

Reply:
Looking at different places it seems that we do the same as stated by @thomasjpfan. I am still not a super fan that we can pass both `1` and `1.0` and that they will have different behaviour.